rollup/internal/controller/relayer/l2_relayer.go

```diff
@@ -729,6 +643,239 @@ func (r *Layer2Relayer) handleL2RollupRelayerConfirmLoop(ctx context.Context) {
+		batch := &encoding.Batch{
+			Index:                      dbBatch.Index,
+			TotalL1MessagePoppedBefore: dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedBefore + dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedInChunk,
+			ParentBatchHash:            parentBatchHash,
+			Chunks:                     chunks,
+		}
+
+		var daBatch *codecv1.DABatch
+		daBatch, err = codecv1.NewDABatch(batch)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to initialize new DA batch: %w", err)
+		}
+
+		encodedChunks := make([][]byte, len(dbChunks))
+		for i, c := range dbChunks {
+			daChunk, createErr := codecv1.NewDAChunk(chunks[i], c.TotalL1MessagesPoppedBefore)
+			if createErr != nil {
+				return nil, nil, fmt.Errorf("failed to initialize new DA chunk: %w", createErr)
+			}
+			encodedChunks[i] = daChunk.Encode()
+		}
+
+		calldata, packErr := r.l1RollupABI.Pack("commitBatch", daBatch.Version, parentBatchHeader, encodedChunks, daBatch.SkippedL1MessageBitmap)
+		if packErr != nil {
+			return nil, nil, fmt.Errorf("failed to pack commitBatch: %w", packErr)
+		}
+		return calldata, daBatch.Blob(), nil
+	}
+
+	// codecv0
+	daBatch, err := codecv0.NewDABatchFromBytes(dbBatch.BatchHeader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to initialize new DA batch from bytes: %w", err)
+	}
+
+	encodedChunks := make([][]byte, len(dbChunks))
+	for i, c := range dbChunks {
+		daChunk, createErr := codecv0.NewDAChunk(chunks[i], c.TotalL1MessagesPoppedBefore)
+		if createErr != nil {
+			return nil, nil, fmt.Errorf("failed to initialize new DA chunk: %w", createErr)
+		}
+		daChunkBytes, encodeErr := daChunk.Encode()
+		if encodeErr != nil {
+			return nil, nil, fmt.Errorf("failed to encode DA chunk: %w", encodeErr)
+		}
+		encodedChunks[i] = daChunkBytes
+	}
+
+	calldata, packErr := r.l1RollupABI.Pack("commitBatch", daBatch.Version, parentBatchHeader, encodedChunks, daBatch.SkippedL1MessageBitmap)
+	if packErr != nil {
+		return nil, nil, fmt.Errorf("failed to pack commitBatch: %w", packErr)
+	}
+	return calldata, nil, nil
+}
+
+func (r *Layer2Relayer) constructFinalizeBatchPayload(dbBatch *orm.Batch, withProof bool) ([]byte, error) {
+	if dbBatch.Index == 0 {
+		return nil, fmt.Errorf("invalid args: batch index is 0, should only happen in finalizing genesis batch")
+	}
+
+	parentDBBatch, getErr := r.batchOrm.GetBatchByIndex(r.ctx, dbBatch.Index-1)
+	if getErr != nil {
+		return nil, fmt.Errorf("failed to get batch, index: %d, err: %w", dbBatch.Index-1, getErr)
+	}
+	parentBatchStateRoot := parentDBBatch.StateRoot
+	parentBatchHash := common.HexToHash(parentDBBatch.Hash)
+
+	dbChunks, err := r.chunkOrm.GetChunksInRange(r.ctx, dbBatch.StartChunkIndex, dbBatch.EndChunkIndex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch chunks: %w", err)
+	}
+
+	startBlockNumber := dbChunks[0].StartBlockNumber
+	if startBlockNumber >= r.banachForkHeight { // codecv1
+		if withProof { // finalizeBatch4844 with proof.
```

Comment:
Can we move the chunk and batch encoding plus the blob data proof outside this if-else?

Reply:
I suggest to make functions like `constructFinalizeBatchPayloadV0` and `constructFinalizeBatchPayloadV1`.

Reply:
good suggestion, will change it into 2 functions.

Reply:
above comments are addressed in [5fcc831](https://github.com/scroll-tech/scroll/pull/1201/commits/5fcc831f512cb8bf2e416f87941a3d79d2dfb28c)